### PR TITLE
Improve the efficiency of crawling github  by making sure a github file is crawled only once

### PR DESCRIPTION
**Pull request by haiyanmeng**
Improve the efficiency of crawling github  by making sure a github file is crawled only once

### CODE REVIEW

File: api/internal/crawl/crawler/crawler.go [hunk 1]
```diff
@@ -29,7 +29,7 @@ type Crawler interface {
 	// Crawl returns when it is done processing. This method does not take
 	// ownership of the channel. The channel is write only, and it
 	// designates where the crawler should forward the documents.
-	Crawl(ctx context.Context, output chan<- CrawledDocument) error
+	Crawl(ctx context.Context, output chan<- CrawledDocument, seen map[string]struct{}) error
```
Comment (monopole) [inline api/internal/crawl/crawler/crawler.go]: can we make `seen` a non-primitive type?
that way we can distinguish it from branchMap more easily.  plus type safety, etc.

```
type seenMap map[string]bool

func (of seenMap) alreadySaw(arg string) bool {
	_, ok := of[arg]
	return ok
}

func (of seenMap) recognize(arg string) {
	of[arg] = true
}

func newSeenMap() seenMap {
	return make(map[string]bool)
}
```

File: api/internal/crawl/crawler/github/crawler.go [hunk 2]
```diff
@@ -176,10 +200,18 @@ func (gc githubCrawler) Match(d *doc.Document) bool {
 	return strings.Contains(repoSpec.Host, "github.com")
 }
 
+type RangeQueryResult struct {
+	totalDocCnt uint64
+	seenDocCnt   uint64
+	newDocCnt uint64
+	errorCnt  uint64
+}
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 3]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
...
-	totalCnt := 0
+	result := RangeQueryResult{}
+	pageID := 1
 	for page := range queryPages {
 		if page.Error != nil {
 			errs = append(errs, page.Error)
 			continue
 		}
-
+		var errorCnt, seenDocCnt, newDocCnt, totalDocCnt uint64
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 4]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
...
 		for _, file := range page.Parsed.Items {
-			k, err := kustomizationResultAdapter(gcl, file)
+			k, err := kustomizationResultAdapter(gcl, file, seen, branchMap)
 			if err != nil {
 				logger.Printf("kustomizationResultAdapter failed: %v", err)
 				errs = append(errs, err)
 				errorCnt++
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 5]
```diff
@@ -79,13 +90,24 @@ func (gc githubCrawler) Crawl(
...
 	for _, query := range ranges {
-		err := processQuery(ctx, gc.client, query, output)
+		result, err := processQuery(ctx, gc.client, query, output, seen, gc.branchMap)
 		if err != nil {
 			errs = append(errs, err)
 		}
+		queryResult.totalDocCnt += result.totalDocCnt
+		queryResult.seenDocCnt += result.seenDocCnt
+		queryResult.newDocCnt += result.newDocCnt
+		queryResult.errorCnt += result.errorCnt
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 6]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
 	}()
 
 	errs := make(multiError, 0)
-	errorCnt := 0
-	totalCnt := 0
+	result := RangeQueryResult{}
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 7]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
...
+			totalDocCnt++
 		}
 
-		logger.Printf("got %d files out of %d from API. %d of %d had errors\n",
-			totalCnt, page.Parsed.TotalCount, errorCnt, totalCnt)
-	}
+		logger.Printf("processQuery [page %d]: got %d files out of %d from API. "+
+			"%d have been seen before. %d are new and sent to the output channel." +
+			"%d have kustomizationResultAdapter errors.",
+			pageID, totalDocCnt, page.Parsed.TotalCount, seenDocCnt, newDocCnt, errorCnt)
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 8]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
...
-}
+		pageID++
+	}
 
-func kustomizationResultAdapter(gcl GhClient, k GhFileSpec) (
-	crawler.CrawledDocument, error) {
+	logger.Printf("Summary of processQuery: got %d files from API. "+
+		"%d have been seen before. %d are new and sent to the output channel." +
+		" %d have kustomizationResultAdapter errors.",
+		result.totalDocCnt, result.seenDocCnt, result.newDocCnt, result.errorCnt)
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

File: api/internal/crawl/crawler/crawler.go (hunk 1 above)
Comment (haiyanmeng) [reply api/internal/crawl/crawler/crawler.go]: Done

File: api/internal/crawl/crawler/github/crawler.go [hunk 9]
```diff
@@ -196,50 +228,75 @@ func processQuery(ctx context.Context, gcl GhClient, query string,
...
 		}
 
-		logger.Printf("got %d files out of %d from API. %d of %d had errors\n",
-			totalCnt, page.Parsed.TotalCount, errorCnt, totalCnt)
-	}
+		logger.Printf("processQuery [page %d]: got %d files out of %d from API. "+
+			"%d have been seen before. %d are new and sent to the output channel." +
+			"%d have kustomizationResultAdapter errors.",
+			pageID, totalDocCnt, page.Parsed.TotalCount, seenDocCnt, newDocCnt, errorCnt)
+		result.totalDocCnt += totalDocCnt
```
[the remark this answers is not in the document]
Comment (haiyanmeng) [reply api/internal/crawl/crawler/github/crawler.go]: Done

**Pull request merged**